File: src/core/database/repositories/entry-status.repository.ts
import type { DatabaseContext } from '../connection.js'
import type { DbEntryStatusRow, EntryStatus } from '../types.js'

export class EntryStatusRepository {
  constructor(private ctx: DatabaseContext) {}

  findByEntryId(entryId: number): EntryStatus | undefined {
    const row = this.ctx.db.prepare('SELECT * FROM entry_status WHERE entry_id = ?').get(entryId) as
      | DbEntryStatusRow
      | undefined

    if (!row) return undefined

    return {
      entryId: row.entry_id,
      isArchived: Boolean(row.is_archived),
      currentVersion: row.current_version,
      updatedAt: new Date(row.updated_at),
    }
  }

  create(entryId: number, currentVersion: number, isArchived = false): void {
    this.ctx.db
      .prepare('INSERT INTO entry_status (entry_id, is_archived, current_version) VALUES (?, ?, ?)')
      .run(entryId, isArchived ? 1 : 0, currentVersion)
  }

  updateCurrentVersion(entryId: number, version: number): void {
    this.ctx.db
      .prepare('UPDATE entry_status SET current_version = ?, updated_at = CURRENT_TIMESTAMP WHERE entry_id = ?')
      .run(version, entryId)
  }

  setArchived(entryId: number, isArchived: boolean): boolean {
    const result = this.ctx.db
      .prepare(`
        UPDATE entry_status
        SET is_archived = ?, updated_at = CURRENT_TIMESTAMP
        WHERE entry_id = ?
      `)
      .run(isArchived ? 1 : 0, entryId)

    return result.changes > 0
  }

  delete(entryId: number): boolean {
    const result = this.ctx.db.prepare('DELETE FROM entry_status WHERE entry_id = ?').run(entryId)
    return result.changes > 0
  }
}

File: internal/database/entry_status_repository.go
package database

import (
	"context"
	"database/sql"
	"fmt"

	sqldb "github.com/vault-md/vaultmd/internal/database/sqlc"
)

type EntryStatusRepository struct {
	ctx *Context
}

func NewEntryStatusRepository(dbCtx *Context) *EntryStatusRepository {
	return &EntryStatusRepository{ctx: dbCtx}
}

func (r *EntryStatusRepository) FindByEntryID(ctx context.Context, entryID int64) (*EntryStatusRecord, error) {
	queries := queriesFromContext(r.ctx)
	if queries == nil {
		return nil, fmt.Errorf("entry status repository: missing database context")
	}

	row, err := queries.FindEntryStatusByEntryID(ctx, entryID)
	if err != nil {
		if err == sql.ErrNoRows {
			return nil, nil
		}
		return nil, err
	}

	record := mapEntryStatusRow(row)
	return &record, nil
}

func (r *EntryStatusRepository) Create(ctx context.Context, entryID, currentVersion int64, isArchived bool) error {
	queries := queriesFromContext(r.ctx)
	if queries == nil {
		return fmt.Errorf("entry status repository: missing database context")
	}

	return queries.InsertEntryStatus(ctx, sqldb.InsertEntryStatusParams{
		EntryID:        entryID,
		IsArchived:     boolToNullInt64(isArchived),
		CurrentVersion: nullInt64(currentVersion),
	})
}

func (r *EntryStatusRepository) UpdateCurrentVersion(ctx context.Context, entryID, version int64) error {
	queries := queriesFromContext(r.ctx)
	if queries == nil {
		return fmt.Errorf("entry status repository: missing database context")
	}

	return queries.UpdateEntryStatusCurrentVersion(ctx, sqldb.UpdateEntryStatusCurrentVersionParams{
		CurrentVersion: nullInt64(version),
		EntryID:        entryID,
	})
}

func (r *EntryStatusRepository) SetArchived(ctx context.Context, entryID int64, archived bool) (bool, error) {
	queries := queriesFromContext(r.ctx)
	if queries == nil {
		return false, fmt.Errorf("entry status repository: missing database context")
	}

	affected, err := queries.UpdateEntryStatusArchived(ctx, sqldb.UpdateEntryStatusArchivedParams{
		IsArchived: boolToNullInt64(archived),
		EntryID:    entryID,
	})
	if err != nil {
		return false, err
	}

	return affected > 0, nil
}

func (r *EntryStatusRepository) Delete(ctx context.Context, entryID int64) (bool, error) {
	queries := queriesFromContext(r.ctx)
	if queries == nil {
		return false, fmt.Errorf("entry status repository: missing database context")
	}

	affected, err := queries.DeleteEntryStatus(ctx, entryID)
	if err != nil {
		return false, err
	}

	return affected > 0, nil
}

func mapEntryStatusRow(row sqldb.EntryStatus) EntryStatusRecord {
	return EntryStatusRecord{
		EntryID:        row.EntryID,
		IsArchived:     optionalBool(row.IsArchived),
		CurrentVersion: optionalInt64(row.CurrentVersion),
		UpdatedAt:      optionalTime(row.UpdatedAt),
	}
}
